common/src/test/java/io/pravega/common/io/serialization/VersionedSerializerTests.java

```diff
@@ -0,0 +1,651 @@
+        SubClass2(int id, int field2) {
+            super(id);
+            this.field2 = field2;
+        }
+
+        static class SubClass2Builder implements ObjectBuilder<SubClass2> {
+        }
+    }
+
+    private static class BaseClassSerializer extends VersionedSerializer.MultiType<BaseClass> {
+        @Override
+        protected void declareSerializers() {
+            serializer(SubClass1.class, 1, new SubClass1Serializer());
+            serializer(SubClass2.class, 2, new SubClass2Serializer());
+            serializer(SubClass11.class, 11, new SubClass11Serializer());
+        }
+    }
+
+    private static class SubClass1Serializer extends VersionedSerializer.WithBuilder<SubClass1, SubClass1.SubClass1Builder> {
+        static final byte VERSION = 0;
+
+        @Override
+        protected SubClass1.SubClass1Builder newBuilder() {
+            return SubClass1.builder1();
+        }
+
+        @Override
+        protected byte writeVersion() {
+            return VERSION;
+        }
+
+        @Override
+        protected void declareVersions() {
+            version(VERSION).revision(0, this::write0, this::read0);
+        }
+
+        private void write0(SubClass1 source, RevisionDataOutput output) throws IOException {
+            if (output.requiresExplicitLength()) {
+                output.length(2 * Integer.BYTES);
+            }
+            output.writeInt(source.id);
+            output.writeInt(source.field1);
+        }
+
+        private void read0(RevisionDataInput input, SubClass1.SubClass1Builder target) throws IOException {
+            target.id(input.readInt());
+            target.field1(input.readInt());
+        }
+    }
+
+    private static class SubClass11Serializer extends VersionedSerializer.WithBuilder<SubClass11, SubClass11.SubClass11Builder> {
+        static final byte VERSION = 0;
+
+        @Override
+        protected SubClass11.SubClass11Builder newBuilder() {
+            return SubClass11.builder11();
+        }
+
+        @Override
+        protected byte writeVersion() {
+            return VERSION;
+        }
+
+        @Override
+        protected void declareVersions() {
+            version(VERSION).revision(0, this::write0, this::read0);
+        }
+
+        private void write0(SubClass11 source, RevisionDataOutput output) throws IOException {
+            if (output.requiresExplicitLength()) {
+                output.length(3 * Integer.BYTES);
+            }
+            output.writeInt(source.id);
+            output.writeInt(source.field1);
+            output.writeInt(source.field11);
+        }
+
+        private void read0(RevisionDataInput input, SubClass11.SubClass11Builder target) throws IOException {
+            target.id(input.readInt());
+            target.field1(input.readInt());
+            target.field11(input.readInt());
+        }
+    }
+
+    private static class SubClass2Serializer extends VersionedSerializer.WithBuilder<SubClass2, SubClass2.SubClass2Builder> {
```

Comment:
do we need to write new revisions on subclasses everytime a revision is added to base class? this could be error prone if it is left to the developer to handle this as revisions could be added to base class without updated revision serializers written for all derived classes.

Reply:
You can choose to invoke `super.serialize` but, as you said, that's error prone and you'll get some weird results. So if you have `B inherit A`, and in `B.Serializer` you invoke `super.deserialize` then that will return an object of type `A`, which you'll need to convert to an object of type `B`, which is cumbersome. Plus this won't work for abstract base types.

If you look in this same file for `SubClass11Serializer` (`SubClass11` is a subclass of `SubClass1` which is a subclass of `BaseClass`), each has its own independent serializer. The parent class serializer is not invoked or even checked in these cases. This is preferred, IMO, since subclasses may add new fields, but their semantics may be very different from the base class. 

In the SegmentStore, I do have a base class for each operation (called `Operation`) which provides a base Serializer. Its sole job is to provide a base for validating a derived class on whether it is ready for serialization; it does not do any actual serialization itself. All Operations share a field (sequence number), and I chose to explicitly serialize the field with each Operation Implementation's serializer instead of trying to cram it into a base serializer, which would be prone to the problems mentioned above.

Reply:
I am talking in terms of say B extends A.
Say we now add new fields to A or replace some field in A. These fields are now implicitly inherited in B. Which means any update to A should warrant update to serialization revisions of B. 
So as a developer, we should be prudent in updating all derived classes of A everytime an update to A happens. 
Ideally if this was implicitly handled by the framework rather than developer having to choose to include it in its revisions which could be error prone.